File: java/lib/src/main/java/algorithm/sorting/InsertionSort.java
package algorithm.sorting;

public class InsertionSort extends Sort {
  @Override
  protected int[] sort(int[] arr, SortType type) {
    int length = arr.length;

    for (int i = 1; i < length; i++) {
      int j = i;
      while (j >= 1 && super.compare(arr[j - 1], arr[j], type)) {
        super.swap(arr, j - 1, j);
        j--;
      }
    }

    return arr;
  }
}

File: kotlin/lib/src/main/kotlin/algorithm/sorting/InsertionSort.kt
package algorithm.sorting

public class InsertionSort : Sort() {
  override fun sort(arr: Array<Int>, type: SortType): Array<Int> {
    val length = arr.size

    for (i in 1..length - 1) {
      var j = i
      while (j >= 1 && compare(arr[j - 1], arr[j], type)) {
        arr.swap(j - 1, j)
        j--
      }
    }

    return arr
  }
}
